node/subsystem-util/src/rolling_session_window.rs

```diff
@@ -0,0 +1,635 @@
+	/// the session of the imported block.
+	///
+	/// this only needs to be called on heads where we are directly notified about import, as sessions do
+	/// not change often and import notifications are expected to be typically increasing in session number.
+	///
+	/// some backwards drift in session index is acceptable.
+	pub async fn cache_session_info_for_head(
+		&mut self,
+		ctx: &mut impl SubsystemContext,
+		block_hash: Hash,
+		block_header: &Header,
+	) -> Result<SessionWindowUpdate, SessionsUnavailable> {
+		if self.window_size == 0 { return Ok(SessionWindowUpdate::Unchanged) }
+
+		let session_index = {
+			let (s_tx, s_rx) = oneshot::channel();
+
+			// The genesis is guaranteed to be at the beginning of the session and its parent state
+			// is non-existent. Therefore if we're at the genesis, we request using its state and
+			// not the parent.
+			ctx.send_message(RuntimeApiMessage::Request(
+				if block_header.number == 0 { block_hash } else { block_header.parent_hash },
+				RuntimeApiRequest::SessionIndexForChild(s_tx),
+			).into()).await;
+
+			match s_rx.await {
+				Ok(Ok(s)) => s,
+				Ok(Err(e)) => return Err(SessionsUnavailable {
+					kind: SessionsUnavailableKind::RuntimeApi(e),
+					info: None,
+				}),
+				Err(e) => return Err(SessionsUnavailable {
+					kind: SessionsUnavailableKind::RuntimeApiUnavailable(e),
+					info: None,
+				}),
+			}
+		};
+
+		match self.earliest_session {
+			None => {
+				// First block processed on start-up.
+
+				let window_start = session_index.saturating_sub(self.window_size - 1);
+
+				match load_all_sessions(ctx, block_hash, window_start, session_index).await {
+					Err(kind) => {
+						Err(SessionsUnavailable {
+							kind,
+							info: Some(SessionsUnavailableInfo {
+								window_start,
+								window_end: session_index,
+								block_hash,
+							}),
+						})
+					},
+					Ok(s) => {
+						let update = SessionWindowUpdate::Initialized {
+							window_start,
+							window_end: session_index,
+						};
+
+						self.earliest_session = Some(window_start);
+						self.session_info = s;
+
+						Ok(update)
+					}
+				}
+			}
+			Some(old_window_start) => {
+				let latest = self.latest_session().expect("latest always exists if earliest does; qed");
+
+				// Either cached or ancient.
+				if session_index <= latest { return Ok(SessionWindowUpdate::Unchanged) }
+
+				let old_window_end = latest;
+
+				let window_start = session_index.saturating_sub(self.window_size - 1);
+
+				// keep some of the old window, if applicable.
+				let overlap_start = window_start.saturating_sub(old_window_start);
+
+				let fresh_start = if latest < window_start {
+					window_start
+				} else {
+					latest + 1
+				};
+
+				match load_all_sessions(ctx, block_hash, fresh_start, session_index).await {
+					Err(kind) => {
+						Err(SessionsUnavailable {
+							kind,
+							info: Some(SessionsUnavailableInfo {
+								window_start: latest +1,
```

Comment:
AFAIU this could also be `window_start` since that's what we wanted the full window to be, but we only started requesting from `fresh_start` and so that's the only data that we're missing.